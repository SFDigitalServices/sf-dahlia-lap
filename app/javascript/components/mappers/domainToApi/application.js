import { toPairs, filter, map, join, compact } from 'lodash'
import { createFieldMapper } from '~/utils/objectUtils'
const checkListToString = (list) => {
  return join(compact(list), ';')
}

const adaPrioritiesMapValues = {
  'vision_impairments': 'Vision impairments',
  'mobility_impairments': 'Mobility impairments',
  'hearing_impairments': 'Hearing impairments'
}

const mapAdaPrioritiesMap = (list) => {
  const selectedPriorities = filter(toPairs(list), ([a, b]) => b)
  return map(selectedPriorities, ([a, b]) => adaPrioritiesMapValues[a])
}

export const applicationFieldMapper = {
  id: 'id',
  primary_applicant_contact: 'primaryApplicantContact',
  application_language: 'applicationLanguage',
  has_military_service: 'hasMilitaryService',
  has_developmental_disability: 'hasDevelopmentalDisability',
  answered_community_screening: 'answeredCommunityScreening',
  housing_voucher_or_subsidy: 'householdVouchersSubsidies',
  terms_acknowledged: 'agreeToTerms',
  number_of_dependents: 'numberOfDependents',
  number_of_seniors: 'numberOfSeniors',
  number_of_minors: 'numberOfMinors',
  reserved_senior: 'hasSenior',
  application_submission_type: 'applicationSubmissionType',
  processing_status: 'processingStatus',
  household_assets: 'householdAssets',
  annual_income: 'annualIncome',
  monthly_income: 'monthlyIncome',
  confirmed_household_annual_income: 'confirmedHouseholdAnnualIncome',
  hh_total_income_with_assets_annual: 'HHTotalIncomeWithAssets',
  total_monthly_rent: 'totalMonthlyRent',
  adaPrioritiesSelected: (source) => checkListToString(mapAdaPrioritiesMap(source.has_ada_priorities_selected)),
  is_first_time_homebuyer: 'isFirstTimeHomebuyer',
  has_completed_homebuyer_education: 'hasCompletedHomebuyerEducation',
  has_loan_preapproval: 'hasLoanPreapproval',
  lending_agent: 'lendingAgent'
}

export const mapApplication = createFieldMapper(applicationFieldMapper)
