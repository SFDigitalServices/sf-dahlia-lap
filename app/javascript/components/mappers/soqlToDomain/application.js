import { split, snakeCase, fromPairs } from 'lodash'
import { mapApplicationMember } from './application_member'
import { mapApplicationPreference } from './application_preference'
import { mapAttachment } from './attachment'
import { mapDemographics } from './demographics'
import { mapFlaggedApplication } from './flagged_application'
import { mapListing } from './listing'
import { mapUser } from './user'
import { mapShape, mapList } from '../utils'
import formUtils from '~/utils/formUtils'

const parseList = text => split(text, ';')
const toChecklist = list => fromPairs(list.map(i => [snakeCase(i), true]))

export const mapApplication = (a) => {
  return {
    applicant: mapShape(mapApplicationMember, a.Applicant),
    demographics: mapShape(mapDemographics, a.Applicant),
    primary_applicant_contact: a.Primary_Applicant,
    alternate_contact: mapShape(mapApplicationMember, a.Alternate_Contact),
    listing: mapShape(mapListing, a.Listing),
    preferences: mapList(mapApplicationPreference, a.preferences),
    proof_files: mapList(mapAttachment, a.proof_files),
    household_members: mapList(mapApplicationMember, a.household_members),
    flagged_applications: mapList(mapFlaggedApplication, a.flagged_applications),
    lease: a.lease,
    rental_assistances: a.rental_assistances,
    createdby: mapShape(mapUser, a.CreatedBy),
    id: a.Id,
    name: a.Name,
    status: a.Status,
    total_household_size: a.Total_Household_Size,
    application_submission_type: a.Application_Submission_Type,
    application_submitted_date: a.Application_Submitted_Date,
    annual_income: formUtils.formatPrice(a.Annual_Income),
    confirmed_household_annual_income: formUtils.formatPrice(a.Confirmed_Household_Annual_Income),
    hh_total_income_with_assets_annual: formUtils.formatPrice(a.HH_Total_Income_with_Assets_Annual),
    household_assets: formUtils.formatPrice(a.Household_Assets),
    monthly_income: a.Monthly_Income,
    housing_voucher_or_subsidy: a.Housing_Voucher_or_Subsidy,
    referral_source: a.Referral_Source,
    application_language: a.Application_Language,
    lottery_number_manual: a.Lottery_Number_Manual,
    lottery_number: a.Lottery_Number,
    total_monthly_rent: formUtils.formatPrice(a.Total_Monthly_Rent),
    general_lottery: a.General_Lottery,
    general_lottery_rank: a.General_Lottery_Rank,
    answered_community_screening: a.Answered_Community_Screening,
    has_military_service: a.Has_Military_Service,
    has_developmental_disability: a.Has_DevelopmentalDisability,
    has_ada_priorities_selected: toChecklist(parseList(a.Has_ADA_Priorities_Selected)),
    terms_acknowledged: a.Terms_Acknowledged,
    number_of_dependents: a.Number_of_Dependents,
    number_of_seniors: a.Number_of_Seniors,
    number_of_minors: a.Number_of_Minors,
    processing_status: a.Processing_Status,
    status_last_updated: a.Status_Last_Updated,
    reserved_senior: a.Reserved_Senior,
    is_first_time_homebuyer: a.Is_First_Time_Homebuyer,
    has_completed_homebuyer_education: a.Has_Completed_Homebuyer_Education,
    has_loan_preapproval: a.Has_Loan_Pre_approval,
    lending_agent: a.Lending_Agent
  }
}
