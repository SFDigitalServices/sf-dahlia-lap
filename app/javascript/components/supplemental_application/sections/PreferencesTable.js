import React from 'react'
import { map, reject, overSome, findIndex, orderBy, kebabCase } from 'lodash'

import TableWrapper from '~/components/atoms/TableWrapper'
import PreferenceIcon from './preferences/PreferenceIcon'
import ExpandableTable from '~/components/molecules/ExpandableTable'
import Panel from './preferences/Panel'
import { isCOP, isDTHP, isAliceGriffith, getPreferenceName } from './preferences/utils'
import { getTypeOfProof } from './preferences/typeOfProof'
import { withContext } from '../context'
import { memberNameFromPref } from '~/components/applications/application_form/preferences/utils'

const { ExpanderButton } = ExpandableTable

const hasExpanderButton = (prefName) => !overSome(isCOP, isDTHP, isAliceGriffith)(prefName)

const onlyValid = (preferences) => {
  return reject(preferences, (pref) => {
    return !pref.receives_preference
  })
}

const matchingPreference = (row) => (preference) => {
  return getPreferenceName(preference) === row[1].content
}

/** Presenter **/

const buildRow = (proofFiles, applicationMembers, fileBaseUrl) => (preference) => {
  return [
    { content: <PreferenceIcon status={preference.post_lottery_validation} /> },
    { content: getPreferenceName(preference) },
    { content: memberNameFromPref(preference.application_member_id, applicationMembers) },
    { content: preference.preference_lottery_rank, classes: ['text-right'] },
    { content: getTypeOfProof(preference, proofFiles, fileBaseUrl) },
    { content: preference.post_lottery_validation }
  ]
}

const buildRows = (application, applicationMembers, fileBaseUrl) => {
  const { preferences } = application
  const proofFiles = application.proof_files
  const sortedPreferences = orderBy(preferences, 'preference_order', 'asc')
  return map(onlyValid(sortedPreferences), buildRow(proofFiles, applicationMembers, fileBaseUrl))
}

const columns = [
  { content: '' },
  { content: 'Preference' },
  { content: 'Claimant' },
  { content: 'Rank', classes: ['text-right'] },
  { content: 'Proof' },
  { content: 'Status' },
  { content: '' }
]

const expandedRowRenderer = (
  application,
  applicationMembers,
  onSave,
  onPanelClose,
  form,
  visited
) => (row, toggle) => {
  const preferenceIndex = findIndex(application.preferences, matchingPreference(row))
  const handleOnClose = (preferenceIndex) => {
    toggle()
    onPanelClose && onPanelClose(preferenceIndex)
  }
  const handleOnSave = async (preferenceIndex, application) => {
    const response = await onSave(preferenceIndex, application)
    response && handleOnClose()
  }

  return (
    <Panel
      application={application}
      applicationMembers={applicationMembers}
      preferenceIndex={preferenceIndex}
      onSave={handleOnSave}
      onClose={handleOnClose}
      form={form}
      visited={visited}
    />
  )
}

const expanderAction = (row, expanded, expandedRowToggler) => {
  const prefName = row[1].content
  return (
    !expanded &&
    hasExpanderButton(prefName) && (
      <ExpanderButton
        label='Edit'
        onClick={expandedRowToggler}
        id={`${kebabCase(prefName)}-edit`}
      />
    )
  )
}

const PreferencesTable = ({
  application,
  applicationMembers,
  fileBaseUrl,
  onSave,
  onPanelClose,
  form,
  visited
}) => {
  const rows = buildRows(application, applicationMembers, fileBaseUrl)
  return (
    <div className='preferences-table'>
      <TableWrapper>
        <ExpandableTable
          columns={columns}
          rows={rows}
          rowKeyIndex={1}
          expanderRenderer={expanderAction}
          expandedRowRenderer={expandedRowRenderer(
            application,
            applicationMembers,
            onSave,
            onPanelClose,
            form,
            visited
          )}
        />
      </TableWrapper>
    </div>
  )
}

export default withContext(PreferencesTable)
