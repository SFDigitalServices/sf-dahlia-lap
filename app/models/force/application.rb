# frozen_string_literal: true

# TODO: Re-enable Rubocop's line length check on this file once
# we have settled on a format for the field mapping items
# rubocop:disable Metrics/LineLength
module Force
  # Represent an application object. Provide mapping between
  # Salesforce object field names, Salesforce custom API field names,
  # and LAP domain field names for applications.
  class Application < Force::ObjectBase
    # TODO: Once we add more models and more fields, consider moving the
    # field mappings into YML files or other places.
    FIELD_NAME_MAPPINGS = [
      { custom_api: 'adaPrioritiesSelected', domain: 'has_ada_priorities_selected', salesforce: 'Has_ADA_Priorities_Selected' },
      { custom_api: 'agreeToTerms', domain: 'terms_acknowledged', salesforce: 'Terms_Acknowledged' },
      { custom_api: 'annualIncome', domain: 'annual_income', salesforce: 'Annual_Income' },
      { custom_api: 'answeredCommunityScreening', domain: 'answered_community_screening', salesforce: 'Answered_Community_Screening' },
      { custom_api: 'applicationLanguage', domain: 'application_language', salesforce: 'Application_Language' },
      { custom_api: 'applicationSubmissionType', domain: 'application_submission_type', salesforce: 'Application_Submission_Type' },
      { custom_api: 'applicationSubmittedDate', domain: 'application_submitted_date', salesforce: 'Application_Submitted_Date' },
      { custom_api: 'appRTType', domain: '', salesforce: '' },
      { custom_api: '', domain: 'createdby', salesforce: 'CreatedBy' },
      { custom_api: 'didApplicantUseHousingCounselingAgency', domain: '', salesforce: 'Applicant_used_housing_counseling_agency' },
      { custom_api: 'externalSessionId', domain: '', salesforce: 'Third_Party_External_ID' },
      { custom_api: 'finalHouseholdIncome', domain: '', salesforce: 'Final_Household_Income' },
      { custom_api: 'formMetadata', domain: '', salesforce: '' },
      { custom_api: '', domain: 'general_lottery', salesforce: 'General_Lottery' },
      { custom_api: '', domain: 'general_lottery_rank', salesforce: 'General_Lottery_Rank' },
      { custom_api: 'hasDevelopmentalDisability', domain: 'has_developmental_disability', salesforce: 'Has_DevelopmentalDisability' },
      { custom_api: 'hasMilitaryService', domain: 'has_military_service', salesforce: 'Has_Military_Service' },
      { custom_api: 'hasPublicHousing', domain: '', salesforce: '' },
      { custom_api: 'hasSenior', domain: 'reserved_senior', salesforce: 'Reserved_Senior' },
      { custom_api: 'HHTotalIncomeWithAssets', domain: 'hh_total_income_with_assets_annual', salesforce: 'HH_Total_Income_with_Assets_Annual' },
      { custom_api: 'householdAssets', domain: 'household_assets', salesforce: 'Household_Assets' },
      { custom_api: 'householdVouchersSubsidies', domain: 'housing_voucher_or_subsidy', salesforce: 'Housing_Voucher_or_Subsidy' },
      { custom_api: 'housingCounselingAgency', domain: '', salesforce: 'Which_One' },
      { custom_api: 'id', domain: 'id', salesforce: 'Id' },
      { custom_api: 'interviewScheduledDate', domain: '', salesforce: 'Interview_Scheduled_Date' },
      { custom_api: '', domain: 'is_lottery_complete', salesforce: 'Is_Lottery_Complete' },
      { custom_api: 'listingID', domain: '', salesforce: 'Listing' },
      { custom_api: 'lotteryNumber', domain: 'lottery_number', salesforce: 'Lottery_Number' },
      { custom_api: 'lotteryNumberManual', domain: 'lottery_number_manual', salesforce: 'Lottery_Number_Manual' },
      { custom_api: 'monthlyIncome', domain: 'monthly_income', salesforce: 'Monthly_Income' },
      { custom_api: 'name', domain: 'name', salesforce: 'Name' },
      { custom_api: 'numberOfDependents', domain: 'number_of_dependents', salesforce: 'Number_of_Dependents' },
      { custom_api: 'numberOfSeniors', domain: 'number_of_seniors', salesforce: 'Number_of_Seniors' },
      { custom_api: 'numberOfMinors', domain: 'number_of_minors', salesforce: 'Number_of_Minors' },
      { custom_api: 'otherHousingCounselingAgency', domain: '', salesforce: 'Other' },
      { custom_api: 'primaryApplicantContact', domain: 'primary_applicant_contact', salesforce: 'Primary_Applicant' },
      { custom_api: 'processingStatus', domain: 'processing_status', salesforce: 'Processing_Status' },
      { custom_api: 'referral', domain: 'referral_source', salesforce: 'Referral_Source' },
      { custom_api: 'snapshotId', domain: '', salesforce: 'Snapshot_ID' },
      { custom_api: 'status', domain: 'status', salesforce: 'Status' },
      { custom_api: '', domain: 'status_last_updated', salesforce: 'Status_Last_Updated' },
      { custom_api: 'subStatus', domain: 'sub_status', salesforce: 'Sub_Status' },
      { custom_api: '', domain: 'total_household_size', salesforce: 'Total_Household_Size' },
      { custom_api: 'totalMonthlyRent', domain: 'total_monthly_rent', salesforce: 'Total_Monthly_Rent' },
    ].freeze

    def to_domain
      domain_fields = super

      # Special field conversion cases for applications
      existing_fields = @fields[:salesforce].presence || @fields[:custom_api].presence
      if existing_fields.present?
        # ADA priorities
        ada_value = existing_fields['Has_ADA_Priorities_Selected'] || existing_fields['adaPrioritiesSelected']
        if ada_value
          ada_arr = ada_value.split(';')
          ada_hash = ada_arr.map { |x| [x.parameterize(separator: '_'), true] }.to_h
          domain_fields['has_ada_priorities_selected'] = ada_hash
        end

        # Created by
        domain_fields['createdby'] = { name: existing_fields['CreatedBy']['Name'] } if existing_fields['CreatedBy']
      end

      domain_fields
    end

    # TODO: There is code somewhere in this app that causes the Salesforce field
    # names we use in SOQL queries to have the "__c" suffix automatically added
    # onto them. That forces the field mappings in these Force objects to use
    # Salesforce field names without that suffix. That in turn causes issues
    # when trying to use the field mappings to translate to Salesforce fields
    # for REST API calls. Find and remove the code that strips out the "__c"
    # suffix, so that the Force objects' field mappings can contain true
    # Salesforce field names. Then remove this function as it won't be needed.
    def to_salesforce
      salesforce_fields = super
      add_salesforce_suffix(salesforce_fields)
    end
  end
end
# rubocop:enable Metrics/LineLength
