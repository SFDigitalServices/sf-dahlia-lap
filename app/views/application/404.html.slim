h1 Record not found
