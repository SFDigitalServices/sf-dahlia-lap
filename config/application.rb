# frozen_string_literal: true

require_relative 'boot'

require 'rails/all'

# Require the gems listed in Gemfile, including any gems
# you've limited to :test, :development, or :production.
Bundler.require(*Rails.groups)

module SfDahliaLap
  # Top level class for LAP application.
  class Application < Rails::Application
    # Initialize configuration defaults for originally generated Rails version.
    config.load_defaults 5.1
    # update defaults when we are done with config/initializers/new_framework_defaults_7_0.rb
    # config.load_defaults 7.0

    # TODO: remove this once we are on Rails 7, only needed as we incrementally upgrade
    config.autoloader = :zeitwerk

    # Disables the deprecated #to_s override in some Ruby core classes
    # See https://guides.rubyonrails.org/configuring.html#config-active-support-disable-to-s-conversion for more information.
    config.active_support.disable_to_s_conversion = true

    # Change the format of the cache entry to 7.0 after deploying the 7.0 upgrade
    config.active_support.cache_format_version = 7.0

    # Rails 7 can protect from open redirect attacks in `redirect_back_or_to` and `redirect_to`.
    # This is not compatible with our authentication process so we disable it
    # config.action_controller.raise_on_open_redirects = false # TODO do we need this?

    # Settings in config/environments/* take precedence over those specified here.
    # Application configuration should go into files in config/initializers
    # -- all .rb files in that directory are automatically loaded.

    config.assets.quiet = true
    config.action_view.logger = nil
  end
end
