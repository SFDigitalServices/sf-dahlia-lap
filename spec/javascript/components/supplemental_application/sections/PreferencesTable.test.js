import React from 'react'
import renderer from 'react-test-renderer'
import PreferencesTable from '~/components/supplemental_application/sections/PreferencesTable'
import application from '../../../fixtures/domain_application'

const applicationMembers = [{id: 'a0n0x000000AbE6AAK', first_name: 'karen', last_name: 'jones'}]
const fileBaseUrl = 'https://test.force.com'
const onSave = () => {}
const onDismissError = () => {}
const formApi = {}

describe('PreferencesTable', () => {
  test('should render a table of only preferences that the application receives (Receives_Preference = true) ', () => {
    // const context = cloneDeep(baseContext)
    const component = renderer.create(
      <PreferencesTable
        application={application}
        applicationMembers={applicationMembers}
        onSave={onSave}
        fileBaseUrl={fileBaseUrl}
        onPanelClose={onDismissError}
        formApi={formApi}
      />
    )

    let tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })
})
