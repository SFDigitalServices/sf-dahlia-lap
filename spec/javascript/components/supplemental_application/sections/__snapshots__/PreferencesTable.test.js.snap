// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PreferencesTable should render a table of only preferences that the application receives (Receives_Preference = true)  1`] = `
<div
  className="preferences-table"
>
  <div
    className="form-grid row expand"
  >
    <div
      className="small-12 column"
    >
      <div
        className="scrollable-table-container-under-xlarge"
      >
        <table
          className="td-light td-plain th-plain"
          role="grid"
        >
          <thead>
            <tr>
              <th
                className=""
                scope="col"
              >
                
              </th>
              <th
                className=""
                scope="col"
              >
                Preference Name
              </th>
              <th
                className=""
                scope="col"
              >
                Person Who Claimed
              </th>
              <th
                className="text-right"
                scope="col"
              >
                Preference Rank
              </th>
              <th
                className=""
                scope="col"
              >
                Type of proof
              </th>
              <th
                className=""
                scope="col"
              >
                Status
              </th>
              <th
                className=""
                scope="col"
              >
                Actions
              </th>
            </tr>
          </thead>
          <tbody>
            <tr
              aria-expanded={false}
              className="tr-expand"
              id="alice-griffith-housing-development-resident-row"
            >
              <td
                className=""
              />
              <td
                className=""
              >
                Alice Griffith Housing Development Resident
              </td>
              <td
                className=""
              >
                karen jones
              </td>
              <td
                className="text-right"
              />
              <td
                className=""
              >
                <ul />
              </td>
              <td
                className=""
              />
              <td />
            </tr>
            <tr
              aria-hidden={true}
              className="tr-expand-content"
              id="alice-griffith-housing-development-resident-panel"
            >
              <td
                className="td-expand-nested no-padding"
                colSpan={7}
              >
                <div
                  className="app-editable expand-wide scrollable-table-nested"
                >
                  <div />
                  <div
                    className="form-grid row"
                  >
                    <div
                      className="form-grid_item column"
                    >
                      <button
                        className="button primary tiny margin-right margin-bottom-none save-panel-btn"
                        disabled={false}
                        onClick={[Function]}
                        type="button"
                      >
                        Save
                      </button>
                      <button
                        className="button secondary tiny margin-bottom-none"
                        disabled={false}
                        onClick={[Function]}
                        type="button"
                      >
                        Cancel
                      </button>
                    </div>
                  </div>
                </div>
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    </div>
  </div>
</div>
`;
